### PR TITLE
Add model for zonal segmentation on prostate MRI

### PR DESCRIPTION
**Pull request by kbressem**
### Description
Add model for zonal segmentation on prostate MRI

### Status
**Ready**

### Please ensure all the checkboxes:

- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [x] In-line docstrings updated.
- [x] Update `version` and `changelog` in `metadata.json`.
- [x] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `mexadata.json`. (I updated to latest stable MONAI version)
- [x] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [x] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.

### CODE REVIEW

File: models/pancreas_ct_dints_segmentation/configs/inference.yaml [hunk 1]
```diff
@@ -1,4 +1,3 @@
----
```
Comment (Nic-Ma) [inline models/pancreas_ct_dints_segmentation/configs/inference.yaml]: I think we don't need to remove this line in the YAML file?
CC @dongyang0122 @wyli .

Thanks.

File: models/prostate_mri_anatomy/configs/inference.json [hunk 2]
```diff
@@ -0,0 +1,153 @@
+{
+    "imports": [
+        "$import pandas as pd",
+        "$import os"
+    ],
+    "bundle_root": "/home/bressekk/Documents/model-zoo/models/prostate_mri_anatomy",
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/configs/inference.json]: Please don't expose your personal path in the bundle, better to use `/workspace/data/prostate_mri_anatomy` instead.

Thanks.

File: models/prostate_mri_anatomy/configs/inference.json [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    "output_dir": "$@bundle_root + '/eval'",
+    "dataset_dir": "/home/bressekk/Desktop/prostate158/prostate158_train/",
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/configs/inference.json]: same comment.

Thanks.

File: models/prostate_mri_anatomy/configs/train.json [hunk 4]
```diff
@@ -0,0 +1,371 @@
+{
+    "imports": [
+        "$import pandas as pd",
+        "$import os",
+        "$import ignite"
+    ],
+    "bundle_root": "/home/bressekk/Documents/model-zoo/models/prostate_mri_anatomy",
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/configs/train.json]: same comment.

Thanks.

File: models/prostate_mri_anatomy/configs/train.json [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+    "ckpt_dir": "$@bundle_root + '/models'",
+    "output_dir": "$@bundle_root + '/eval'",
+    "dataset_dir": "/home/bressekk/Desktop/prostate158/prostate158_train/",
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/configs/train.json]: same comment.

Thanks.

File: models/prostate_mri_anatomy/configs/train.json [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+    },
+    "network": "$@network_def.to(@device)",
+    "loss": {
+        "_target_": "DiceFocalLoss",
+        "to_onehot_y": true,
+        "softmax": true,
+        "include_background": false
+    },
+    "optimizer": {
+        "_target_": "monai.optimizers.Novograd",
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/configs/train.json]: You can use `Novograd` directly, no need to full module path.

Thanks

File: models/prostate_mri_anatomy/configs/inference.json [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                "resample": false,
+                "meta_keys": "pred_meta_dict",
+                "output_dir": "@output_dir"
+            }
+        ]
+    },
+    "handlers": [
+        {
+            "_target_": "CheckpointLoader",
+            "load_path": "$@bundle_root + '/models/anatomy.pt'",
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/configs/inference.json]: I think the saved model name in your `train.json` is `model.pt`.
Please update this line to be consistent.

Thanks.

File: models/prostate_mri_anatomy/large-files.yaml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+large_files:
+  - path: "models/anatomy.pt"
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/large-files.yaml]: Please use `model.pt` as the name, we may consume this model in other applications (MONAI label, MONAI deploy, etc.), better to have a common file name.

Thanks.

File: models/prostate_mri_anatomy/scripts/center_crop.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
+import argparse
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/scripts/center_crop.py:13]: Hi @wyli ,

Do we need any copyright in the python scripts of bundle?
I am not very sure about it, what do you think?

Thanks.

File: models/prostate_mri_anatomy/scripts/center_crop.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+from typing import Union
+
+import SimpleITK as sitk  # noqa N813
```
Comment (Nic-Ma) [inline models/prostate_mri_anatomy/scripts/center_crop.py:16]: I think we should add it to the optional packages of `metadata.json` even it's not used in the config files.
@wyli @ericspod What do you think?

Thanks.

Comment (wyli) [reply models/prostate_mri_anatomy/scripts/center_crop.py:16]: the cropping step is optional and I think it's ok to not include it in the metadata.

Comment (Nic-Ma) [reply models/prostate_mri_anatomy/scripts/center_crop.py:16]: OK, sounds good to me, I didn't notify it's optional.

File: models/prostate_mri_anatomy/configs/inference.json (hunk 2 above)
Comment (kbressem) [reply models/prostate_mri_anatomy/configs/inference.json]: You are right. I should have adapted this before the PR. I now use:
"/workspace/data/prostate_mri_anatomy" as bundle root and 
"/workspace/data" as data root.

File: models/prostate_mri_anatomy/configs/inference.json (hunk 3 above)
Comment (kbressem) [reply models/prostate_mri_anatomy/configs/inference.json]: I've adapted the filepath to "/workspace/data/prostate158/prostate158_train/" as "prostate158/prostate158_train/" resembles the folder structure in the prostate158 dataset.

File: models/prostate_mri_anatomy/configs/train.json [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+            "amp": true
+        }
+    },
+    "validate": {
+        "preprocessing": {
+            "_target_": "Compose",
+            "transforms": "%train#deterministic_transforms"
+        },
+        "dataset": {
+            "_target_": "CacheDataset",
```
Comment (kbressem) [inline models/prostate_mri_anatomy/configs/train.json]: I just realized, that this should also use PersistentDataset and images from `valid.csv`. The config was updated accordingly.

File: models/prostate_mri_anatomy/configs/inference.json (hunk 7 above)
Comment (kbressem) [reply models/prostate_mri_anatomy/configs/inference.json]: The model names are now consistent in all files. It's always `model.pt`

File: models/pancreas_ct_dints_segmentation/configs/inference.yaml (hunk 1 above)
Comment (kbressem) [reply models/pancreas_ct_dints_segmentation/configs/inference.yaml]: Sorry. It was shown as an error to me in VSCode. I'll revert the commit.

File: models/prostate_mri_anatomy/scripts/center_crop.py (hunk 9 above)
Comment (kbressem) [reply models/prostate_mri_anatomy/scripts/center_crop.py:13]: Some other scripts also miss a copyright. E.g., 

https://github.com/Project-MONAI/model-zoo/blob/dev/models/brats_mri_segmentation/scripts/prepare_datalist.py
https://github.com/Project-MONAI/model-zoo/blob/dev/models/mednist_gan/scripts/losses.py
https://github.com/Project-MONAI/model-zoo/blob/dev/models/pancreas_ct_dints_segmentation/scripts/prepare_datalist.py

However, some have a copyright: 
https://github.com/Project-MONAI/model-zoo/blob/dev/models/pancreas_ct_dints_segmentation/scripts/search.py
https://github.com/Project-MONAI/model-zoo/blob/dev/models/wholeBrainSeg_Large_UNEST_segmentation/scripts/networks/nest_transformer_3D.py
https://github.com/Project-MONAI/model-zoo/blob/dev/models/wholeBrainSeg_Large_UNEST_segmentation/scripts/networks/patchEmbed3D.py

I've added a Copyright to the file and can also add one to the other scripts in a new PR (if desired)

Comment (ericspod) [reply models/prostate_mri_anatomy/scripts/center_crop.py:13]: One of these files is in the mednist_gan bundle I put up, it's such a tiny amount of trivial code I'm not worried about its license status.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update.
Looks good to me now.
@yiheng-wang-nv @wyli Do you have any other comments?

Thanks.

**Review by wyli — APPROVED** [review]

**Comment by wyli** [comment]
could you please help look into the testing issues? @Nic-Ma @yiheng-wang-nv it's taking 5+ days to merge it...

**Comment by Nic-Ma** [comment]
Hi @kbressem ,

For future deployment and application, could you please help also provide the `TorchScript` model?
You can easily execute the command to export TorchScript model:
```python -m monai.bundle ckpt_export --net_id network_def --filepath models/verify_model.ts --ckpt_file models/model.pt --meta_file configs/metadata.json --config_file configs/inference.json```
And put it in the public storage, referring to:
Refer to: https://github.com/Project-MONAI/model-zoo/blob/dev/models/spleen_ct_segmentation/large_files.yml#L6

Thanks in advance.

**Comment by wyli** [comment]
> Hi @kbressem ,
> 
> For future deployment and application, could you please help also provide the `TorchScript` model? You can easily execute the command to export TorchScript model: `python -m monai.bundle ckpt_export --net_id network_def --filepath models/verify_model.ts --ckpt_file models/model.pt --meta_file configs/metadata.json --config_file configs/inference.json` And put it in the public storage, referring to: Refer to: https://github.com/Project-MONAI/model-zoo/blob/dev/models/spleen_ct_segmentation/large_files.yml#L6
> 
> Thanks in advance.

I just tested it works fine with the `ckpt_export` command.

**Comment by kbressem** [comment]
@wyli this means I don't need to provide the `TorchScript` model. I can still upload it to Zenodo and update the `large_files.yaml` in a small future PR.

**Comment by wyli** [comment]
yes, I think so, the PR looks good to me and it passes all tests.
